File: c++/task1.cpp

#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <functional>

using namespace std;

bool addProduct(string nameProduct, int countProduct, string cellAddress, map<char, vector<vector<vector<map<string, int>>>>> &stock);
int quantityShelfOccupancy(map<char, vector<vector<vector<map<string, int>>>>> &stock, char storageArea, int numberRack, int numberVerticalSection, int shelfNumber);

bool removeProduct(string nameProduct, int countProduct, string cellAddress, map<char, vector<vector<vector<map<string, int>>>>> &stock);
void REM(map<char, vector<vector<vector<map<string, int>>>>> &stock);

void INFO(const map<char, vector<vector<vector<map<string, int>>>>> &stock);
void ADD(map<char, vector<vector<vector<map<string, int>>>>> &stock);

/*
2-зоны хранения - ключ к 1 map
4-стелажа - первый vector
18-вертикальных секций - второй вектор
5-количество полок - 3 вектор
7200 - единиц вместимости
*/
#define MAX_CELL_SIZE 10
#define TOTAL_NUMBER_RACK 4
#define TOTAL_NUMBER_VERTICAL_SELECTION 18
#define TOTAL_SHELF_NUMBER 5
#define TOTAL_PRODUCT_STOCK 7200

int main()
{
    map<char, vector<vector<vector<map<string, int>>>>> stock;

    char action;

    cout << "A = ADD R = REMOVE I = INFO Q = exit" << endl;

    while (true)
    {
        cout << endl;
        cout << "Enter action: ";
        cin >> action;
        switch (action)
        {
        case 'A':
            ADD(stock);
            break;
        case 'R':
            REM(stock);
            break;
        case 'I':
            INFO(stock);
            cout << endl;
            break;
        case 'Q':
            return 0;
        default:
            cout << "unknown team\n";
        }
    }
}

bool addProduct(string nameProduct, int countProduct, string cellAddress, map<char, vector<vector<vector<map<string, int>>>>> &stock)
{
    if (countProduct > 10)
    {
        cout << "cell can only store 10 units\n";
        return false;
    }

    char storageArea = cellAddress[0];
    int numberRack = stoi(cellAddress.substr(1, 2)) - 1;
    int numberVerticalSection = stoi(cellAddress.substr(3, 2)) - 1;
    int shelfNumber = stoi(cellAddress.substr(5, 6)) - 1;

    // ограничение склада
    if (storageArea != 'A' && storageArea != 'B' || numberRack < 0 || numberRack >= TOTAL_NUMBER_RACK || numberVerticalSection < 0 || numberVerticalSection >= TOTAL_NUMBER_VERTICAL_SELECTION || shelfNumber < 0 || shelfNumber >= TOTAL_SHELF_NUMBER)
    {
        cout << "Invalid cell address" << endl;
        return false;
    }

    if (stock.count(storageArea) == 0)
    {
        stock[storageArea] = vector<vector<vector<map<string, int>>>>(TOTAL_NUMBER_RACK, vector<vector<map<string, int>>>(TOTAL_NUMBER_VERTICAL_SELECTION, vector<map<string, int>>(TOTAL_SHELF_NUMBER)));
    }

    // смотрим сколько едениц продуктов на полке
    int numberProductUnitsShelf = quantityShelfOccupancy(stock, storageArea, numberRack, numberVerticalSection, shelfNumber);

    if ((numberProductUnitsShelf + countProduct) <= MAX_CELL_SIZE)
    {
        auto &innerMap = stock[storageArea][numberRack][numberVerticalSection][shelfNumber];
        innerMap[nameProduct] += countProduct;
        return true;
    }
    else
    {
        cout << "now on the shelf " << numberProductUnitsShelf << endl;
        cout << "total shelf can store " << MAX_CELL_SIZE << " units of product" << endl;
        return false;
    }
}

void INFO(const map<char, vector<vector<vector<map<string, int>>>>> &stock)
{
    if (stock.empty())
    {
        cout << "Stock empty" << endl;
        return;
    }

    float counterProduct = 0;
    float counterProductSectorA = 0;
    float counterProductSectorB = 0;

    for (auto &map1 : stock)
    {
        for (size_t vector1 = 0; vector1 < map1.second.size(); vector1++)
        {
            for (size_t vector2 = 0; vector2 < map1.second[vector1].size(); vector2++)
            {
                for (size_t vector3 = 0; vector3 < map1.second[vector1][vector2].size(); vector3++)
                {
                    for (auto &map2 : map1.second[vector1][vector2][vector3])
                    {
                        counterProduct += map2.second;
                        if (map1.first == 'A')
                        {
                            counterProductSectorA += map2.second;
                        }
                        else if (map1.first == 'B')
                        {
                            counterProductSectorB += map2.second;
                        }

                        if (map2.second != 0)
                        {
                            cout << map1.first << "-" << vector1 + 1 << "-" << vector2 + 1 << "-" << vector3 + 1 << "-" << map2.first << " = " << map2.second << endl;
                        }
                    }
                }
            }
        }
        if (map1.first == 'A')
        {
            cout << (counterProductSectorA * 100) / (TOTAL_PRODUCT_STOCK / 2) << "% - sector A" << endl;
        }
        else if (map1.first == 'B')
        {
            cout << (counterProductSectorB * 100) / (TOTAL_PRODUCT_STOCK / 2) << "% - sector B" << endl;
        }
        cout << endl;
    }
    cout << endl;
    cout << (counterProduct * 100) / TOTAL_PRODUCT_STOCK << "% - fullness";
}

void ADD(map<char, vector<vector<vector<map<string, int>>>>> &stock)
{
    string nameProduct;
    int countProduct;
    string cellAddress;

    cout << "nameProduct: ";
    cout << "countProduct: ";
    cout << "cellAddress: ";
    cin >> nameProduct >> countProduct >> cellAddress;
    if (cellAddress.size() != 7)
    {
        cout << "error";
        return;
    }

    if (addProduct(nameProduct, countProduct, cellAddress, stock))
    {
        cout << "product added\n";
    }
    else
    {
        cout << "product not added\n";
    }
}

int quantityShelfOccupancy(map<char, vector<vector<vector<map<string, int>>>>> &stock, char storageArea, int numberRack, int numberVerticalSection, int shelfNumber)
{
    int numberProductUnitsShelf = 0;

    if (stock.count(storageArea) > 0 && numberRack >= 0 && numberRack < TOTAL_NUMBER_RACK && numberVerticalSection >= 0 && numberVerticalSection < TOTAL_NUMBER_VERTICAL_SELECTION && shelfNumber >= 0 && shelfNumber < TOTAL_SHELF_NUMBER)
    {
        auto &innerMap = stock[storageArea][numberRack][numberVerticalSection][shelfNumber];

        for (auto &product : innerMap)
        {
            numberProductUnitsShelf += product.second;
        }
    }

    return numberProductUnitsShelf;
}

bool removeProduct(string nameProduct, int countProduct, string cellAddress, map<char, vector<vector<vector<map<string, int>>>>> &stock)
{
    if (countProduct > 10)
    {
        cout << "No more than 10 can be deleted at a time\n";
        return false;
    }

    char storageArea = cellAddress[0];
    int numberRack = stoi(cellAddress.substr(1, 2)) - 1;
    int numberVerticalSection = stoi(cellAddress.substr(3, 2)) - 1;
    int shelfNumber = stoi(cellAddress.substr(5, 6)) - 1;

    // ограничение склада
    if (storageArea != 'A' && storageArea != 'B' || numberRack < 0 || numberRack >= TOTAL_NUMBER_RACK || numberVerticalSection < 0 || numberVerticalSection >= TOTAL_NUMBER_VERTICAL_SELECTION || shelfNumber < 0 || shelfNumber >= TOTAL_SHELF_NUMBER)
    {
        cout << "Invalid cell address" << endl;
        return false;
    }

    if (stock.empty())
    {
        cout << "Stock empty";
        return 0;
    }

    auto &innerMap = stock[storageArea][numberRack][numberVerticalSection][shelfNumber];

    if (innerMap.find(nameProduct) != innerMap.end())
    {
        if ((innerMap[nameProduct] - countProduct) < 0)
        {
            cout << nameProduct << " is less of this product than you want to remove" << endl;
            return 0;
        }
        else
        {
            innerMap[nameProduct] -= countProduct;

            if (innerMap[nameProduct] == 0)
            {
                innerMap.erase(nameProduct);
            }

            return 1;
        }
    }
    else
    {
        cout << "product not finded" << endl;
        return 0;
    }
}

void REM(map<char, vector<vector<vector<map<string, int>>>>> &stock)
{
    string nameProduct;
    int countProduct;
    string cellAddress;

    cout << "nameProduct: ";
    cout << "countProduct: ";
    cout << "cellAddress: ";
    cin >> nameProduct >> countProduct >> cellAddress;

    if (removeProduct(nameProduct, countProduct, cellAddress, stock))
    {
        cout << "product remove\n";
    }
    else
    {
        cout << "product not remove\n";
    }
}

File: rust/laba5/src/task1.rs
use std::collections::HashMap;
use std::io::{self, Write};

const MAX_CELL_SIZE: usize = 10;
const TOTAL_NUMBER_RACK: usize = 4;
const TOTAL_NUMBER_VERTICAL_SELECTION: usize = 18;
const TOTAL_SHELF_NUMBER: usize = 5;
const TOTAL_PRODUCT_STOCK: usize = 7200;

type Stock = HashMap<char, Vec<Vec<Vec<HashMap<String, i32>>>>>;

fn add_product(
    name_product: &str,
    count_product: i32,
    cell_address: &str,
    stock: &mut Stock,
) -> bool {
    if count_product > 10 {
        println!("cell can only store 10 units");
        return false;
    }

    let storage_area = cell_address.chars().next().unwrap();
    let number_rack = cell_address[1..3].parse::<usize>().unwrap() - 1;
    let number_vertical_section = cell_address[3..5].parse::<usize>().unwrap() - 1;
    let shelf_number = cell_address[5..].parse::<usize>().unwrap() - 1;

    if (storage_area != 'A' && storage_area != 'B')
        || number_rack >= TOTAL_NUMBER_RACK
        || number_vertical_section >= TOTAL_NUMBER_VERTICAL_SELECTION
        || shelf_number >= TOTAL_SHELF_NUMBER
    {
        println!("Invalid cell address");
        return false;
    }

    let stock_entry = stock
        .entry(storage_area)
        .or_insert_with(|| {
            vec![
                vec![
                    HashMap::new();
                    TOTAL_NUMBER_VERTICAL_SELECTION
                ];
                TOTAL_NUMBER_RACK
            ]
        });

    let number_product_units_shelf = quantity_shelf_occupancy(stock, storage_area, number_rack, number_vertical_section, shelf_number);

    if number_product_units_shelf + count_product <= MAX_CELL_SIZE {
        stock_entry[number_rack][number_vertical_section][shelf_number]
            .entry(name_product.to_string())
            .and_modify(|count| *count += count_product)
            .or_insert(count_product);
        true
    } else {
        println!(
            "now on the shelf {} \ntotal shelf can store {} units of product",
            number_product_units_shelf, MAX_CELL_SIZE
        );
        false
    }
}

fn quantity_shelf_occupancy(
    stock: &Stock,
    storage_area: char,
    number_rack: usize,
    number_vertical_section: usize,
    shelf_number: usize,
) -> i32 {
    let mut number_product_units_shelf = 0;

    if let Some(inner_map) = stock.get(&storage_area).and_then(|v| v.get(number_rack)).and_then(|v| v.get(number_vertical_section)).and_then(|v| v.get(shelf_number)) {
        for (_, count) in inner_map {
            number_product_units_shelf += count;
        }
    }

    number_product_units_shelf
}

fn remove_product(
    name_product: &str,
    count_product: i32,
    cell_address: &str,
    stock: &mut Stock,
) -> bool {
    if count_product > 10 {
        println!("No more than 10 can be deleted at a time");
        return false;
    }

    let storage_area = cell_address.chars().next().unwrap();
    let number_rack = cell_address[1..3].parse::<usize>().unwrap() - 1;
    let number_vertical_section = cell_address[3..5].parse::<usize>().unwrap() - 1;
    let shelf_number = cell_address[5..].parse::<usize>().unwrap() - 1;

    if (storage_area != 'A' && storage_area != 'B')
        || number_rack >= TOTAL_NUMBER_RACK
        || number_vertical_section >= TOTAL_NUMBER_VERTICAL_SELECTION
        || shelf_number >= TOTAL_SHELF_NUMBER
    {
        println!("Invalid cell address");
        return false;
    }

    if let Some(inner_map) = stock.get_mut(&storage_area).and_then(|v| v.get_mut(number_rack)).and_then(|v| v.get_mut(number_vertical_section)).and_then(|v| v.get_mut(shelf_number)) {
        if let Some(count) = inner_map.get_mut(name_product) {
            if *count - count_product < 0 {
                println!("{} is less of this product than you want to remove", name_product);
                return false;
            } else {
                *count -= count_product;
                if *count == 0 {
                    inner_map.remove(name_product);
                }
                true
            }
        } else {
            println!("product not found");
            false
        }
    } else {
        println!("Stock empty");
        false
    }
}

fn info(stock: &Stock) {
    if stock.is_empty() {
        println!("Stock empty");
        return;
    }

    let mut counter_product = 0.0;
    let mut counter_product_sector_a = 0.0;
    let mut counter_product_sector_b = 0.0;

    for (storage_area, racks) in stock {
        for rack in racks {
            for vertical_section in rack {
                for shelf in vertical_section {
                    for (name_product, count) in shelf {
                        counter_product += *count as f32;
                        if *storage_area == 'A' {
                            counter_product_sector_a += *count as f32;
                        } else if *storage_area == 'B' {
                            counter_product_sector_b += *count as f32;
                        }

                        if *count != 0 {
                            println!(
                                "{}-{}-{}-{}-{} = {}",
                                storage_area, rack.index() + 1, vertical_section.index() + 1, shelf.index() + 1, name_product, count
                            );
                        }
                    }
                }
            }
        }

        if *storage_area == 'A' {
            println!(
                "{:.2}% - sector A",
                (counter_product_sector_a * 100.0) / (TOTAL_PRODUCT_STOCK as f32 / 2.0)
            );
        } else if *storage_area == 'B' {
            println!(
                "{:.2}% - sector B",
                (counter_product_sector_b * 100.0) / (TOTAL_PRODUCT_STOCK as f32 / 2.0)
            );
        }
        println!();
    }

    println!();
    println!("{:.2}% - fullness", (counter_product * 100.0) / TOTAL_PRODUCT_STOCK as f32);
}

fn add(stock: &mut Stock) {
    let mut name_product = String::new();
    let mut count_product: i32 = 0;
    let mut cell_address = String::new();

    print!("nameProduct: ");
    io::stdout().flush().unwrap();
    io::stdin().read_line(&mut name_product).unwrap();
    name_product = name_product.trim().to_string();

    print!("countProduct: ");
    io::stdout().flush().unwrap();
    io::stdin().read_line(&mut count_product).unwrap();
    count
